device/cuda/src/seeding/seed_finding.cu

```diff
@@ -155,158 +165,167 @@ seed_finding::output_type seed_finding::operator()(
         make_prefix_sum_buff(grid_sizes, *m_copy, m_mr);
 
     // Set up the doublet counter buffer.
-    device::doublet_counter_container_types::buffer doublet_counter_buffer =
-        device::make_doublet_counter_buffer(grid_sizes, *m_copy, m_mr.main,
-                                            m_mr.host);
+    device::doublet_counter_collection_types::buffer doublet_counter_buffer = {
+        m_copy->get_size(sp_grid_prefix_sum_buff), 0, m_mr.main};
+    m_copy->setup(doublet_counter_buffer);
 
     // Calculate the number of threads and thread blocks to run the doublet
     // counting kernel for.
     const unsigned int nDoubletCountThreads = WARP_SIZE * 2;
     const unsigned int nDoubletCountBlocks =
-        (sp_grid_prefix_sum_buff.size() + nDoubletCountThreads - 1) /
+        (m_copy->get_size(sp_grid_prefix_sum_buff) + nDoubletCountThreads - 1) /
         nDoubletCountThreads;
 
+    // Counters for the total number of doublets
+    vecmem::unique_alloc_ptr<unsigned int> num_midBot_device =
+        vecmem::make_unique_alloc<unsigned int>(m_mr.main);
+    vecmem::unique_alloc_ptr<unsigned int> num_midTop_device =
+        vecmem::make_unique_alloc<unsigned int>(m_mr.main);
+    CUDA_ERROR_CHECK(
+        cudaMemset(num_midBot_device.get(), 0, sizeof(unsigned int)));
+    CUDA_ERROR_CHECK(
+        cudaMemset(num_midTop_device.get(), 0, sizeof(unsigned int)));
```

Comment:
It's a small thing, but we could use a dedicated type just for these two counters as well. That way we could halve the number of allocations/de-allocations necessary to handle these.

Or in fact, you could declare a `struct` for **all** of the `unsigned int` variables needed by the algorithm, and allocate/de-allocate them all at once. That `struct` should probably go under `traccc/edm/device` of course.